cryptolib/verifier.go

```diff
@@ -0,0 +1,59 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cryptolib
+
+import "errors"
+
+// Verifier contains methods to validate an Attestation.
+type Verifier interface {
+	VerifyAttestation(att *Attestation, publicKey []byte, image string) error
+}
+
+type verifier struct{}
+
+// NewVerifier creates a Verifier interface.
+func NewVerifier() Verifier {
+	return &verifier{}
+}
+
+// VerifyAttestation verifies whether an Attestation satisfies an
+// AttestationAuthority's public key under an image.
+func (v *verifier) VerifyAttestation(att *Attestation, publicKey []byte, image string) error {
```

Comment:
Public key sometimes comes with an id (PGP), but sometimes doesn't (both PKIX and JWT, I think).  During verification, we will require that key IDs match, not just that the signature verifies.

So -- while key ID matching could happen outside of this function, at the calling site, I'd consider pulling it in here explicitly.  This would require adding a key id argument in some form.

It could be an explicit argument, and then the relationship between key id and publicKey should be clarified (e.g. for PGP).  You could also create a dedicated PublicKey interface which encapsulates both the id and the key material.

Potential added benefit of a dedicated PublicKey interface is that it further decouples VerifyAttestation from wire format.  But that's a bit unclear still until you document how publicKey argument should be set.

Reply:
+1 to a dedicated PublicKey interface.  That will make it easy to enforce that any attestation-provided signing algorithm isn't accidentally trusted for the purpose of signature verification.  It should also make testing easier.

It's also easy to imagine a PublicKeySet with the same interface that encapsulates a bundle of PublicKey instances and returns iff at least one of the encapsulated PublicKeys verifies the attestation.  It could also be responsible for dispatching to the right PublicKey based on keyId, and rejecting if it doesn't know about any such keyId.

Reply:
+1 to what's said above, 1) the crypto library should also include the logic of public key match-up by key id, and 2) once key is looked up, we only need to work on the crypto level, without worrying about high-level concepts like attestation.

Reply:
We have encapsulated the key data and id in a `PublicKey` struct. We also modified our `VerifyAttestation` function to receive `publicKeySet`, a list of `PublicKeys`.